### PR TITLE
miniBUDE Timer Updates

### PR DESCRIPTION
**Pull request by pranav-sivaraman**
- [x] Double check data movement
- [x] Add configurable warmup iterations

### CODE REVIEW

File: src/acc/fasten.hpp [hunk 1]
```diff
@@ -158,78 +158,77 @@ template <size_t PPWI> class IMPL_CLS final : public Bude<PPWI> {
...
+    delete(protein, ligand, forcefield,       \
+           poses_0, poses_1, poses_2,         \
+           poses_3, poses_4, poses_5,         \
+           energies)
+
     std::copy(energies, energies + p.nposes(), sample.energies.begin());
-    std::free(protein);
-    std::free(ligand);
-    std::free(forcefield);
-    std::free(energies);
```
Comment (jhdavis8) [inline src/acc/fasten.hpp:230]: `energies` is calloc'd above so I believe we need to free it before the name goes out of scope.

File: src/omp/fasten.hpp [hunk 2]
```diff
@@ -253,13 +228,19 @@ template <size_t PPWI> class IMPL_CLS final : public Bude<PPWI> {
       sample.kernelTimes.emplace_back(kernelStart, kernelEnd);
     }
 
+    auto deviceToHostStart = now();
+  #pragma omp target update from(energies[:nposes])
     std::copy(energies, energies + p.nposes(), sample.energies.begin());
-    std::free(protein);
-    std::free(ligand);
-    std::free(forcefield);
-    std::free(energies);
```
Comment (jhdavis8) [inline src/omp/fasten.hpp:260]: Same here for energies, I think it still needs to be freed like for OpenACC

**Review by jhdavis8 — CHANGES_REQUESTED** [review]
Looks good, two small changes to discuss

File: src/acc/fasten.hpp (hunk 1 above)
Comment (jhdavis8) [reply src/acc/fasten.hpp:230]: Or change energies to point to sample.energies

File: src/acc/fasten.hpp [hunk 3]
```diff
@@ -158,78 +158,75 @@ template <size_t PPWI> class IMPL_CLS final : public Bude<PPWI> {
...
-    for (auto &pose : poses)
-      std::free(pose);
+
+  auto deviceToHostStart = now();
+        
+  #pragma acc update                          \
+    host(energies[:nposes])
+        
+  auto deviceToHostEnd = now();
+  sample.deviceToHost = {deviceToHostStart, deviceToHostEnd};
```
Comment (joy-kitson) [inline src/acc/fasten.hpp:225]: Is there a reason we're saving the start and end times rather than just the time elapsed?

File: src/bude.h [hunk 4]
```diff
@@ -85,9 +86,10 @@ struct Sample {
   size_t ppwi, wgsize;
   std::vector<float> energies;
   std::vector<std::pair<TimePoint, TimePoint>> kernelTimes;
-  std::optional<std::pair<TimePoint, TimePoint>> contextTime;
+  std::optional<std::pair<TimePoint, TimePoint>> hostToDevice;
+  std::optional<std::pair<TimePoint, TimePoint>> deviceToHost;
```
Comment (joy-kitson) [inline src/bude.h:91]: I guess it makes sense to do it this way if that's what's in the existing code, but it does still seem rather unorthodox, and also doesn't quite line up with what we do in the other codes

File: src/cuda/fasten.hpp [hunk 5]
```diff
@@ -253,7 +253,7 @@ template <size_t PPWI> class IMPL_CLS final : public Bude<PPWI> {
 
     Sample sample(PPWI, wgsize, p.nposes());
 
-    auto contextStart = now();
+    auto hostToDeviceStart = now();
```
Comment (joy-kitson) [inline src/cuda/fasten.hpp:256]: Does the `allocate` function preform a copy for the passed buffer? If not, I thought we weren't planning on timing memory allocations, just data movement?

Comment (joy-kitson) [reply src/cuda/fasten.hpp:256]: Took a look at it and it seems to just allocate, so not sure if you intend to time those (if you do I probably need to adjust some timers in Cloverleaf)

**Review by jhdavis8 — APPROVED** [review]

**Comment by jhdavis8** [comment]
@joy-kitson would you be able to review this one as well?

**Review by joy-kitson — COMMENTED** [review]
It looks like in most of the models you're timing allocation under the umbrella of host to device data movement, unless I'm missing something. I think it might be worth discussing whether or not this is the intended behavior for our timers, as if it is I think I need to make some changes in Cloverleaf (and if memory serves, we may need to do so in some of the other apps as well).

File: src/bude.h (hunk 4 above)
Comment (pranav-sivaraman) [reply src/bude.h:91]: The final output is just the elapsed time so I think it shouldn't be a problem copying the way they do it.

File: src/acc/fasten.hpp (hunk 3 above)
Comment (pranav-sivaraman) [reply src/acc/fasten.hpp:225]: See below

File: src/cuda/fasten.hpp (hunk 5 above)
Comment (pranav-sivaraman) [reply src/cuda/fasten.hpp:256]: `allocate` is an overloaded function where the argument is a `std::vector` which then calls `allocate` again where the argument is just a size for the allocation. 

But we decided in the meeting that we are including the allocation time and copy so this is a non-issue tho I need to ensure that is consistent

**Pull request merged**